Fix update-checkout error reporting

etcwilde opened the pull request:
In the event of a failure, update-checkout wouldn't always fail very
gracefully. As a result the actual failure wasn't always printed, which
makes it really difficult to figure out what's going on.

In my case, I'm jumping between 5.6 and main branches, so
the `swift-cmark-gfm` is causing update-checkout to hiccup since it's
not in the `main` configuration file scheme for `main`, but is in the
5.6 repository. You wouldn't know that though, since it would complain
about `repo_path` not being in the error message before crashing.

I've gone ahead and done some cleanups, reporting what the problem is at
the point of failure beyond the "key error 'swift-cmark-cfg'.

etcwilde committed:
Fix update-checkout error reporting

In the event of a failure, update-checkout wouldn't always fail very
gracefully. As a result the actual failure wasn't always printed, which
makes it really difficult to figure out what's going on.

In my case, I'm jumping between 5.6 and main branches, so
the `swift-cmark-gfm` is causing update-checkout to hiccup since it's
not in the `main` configuration file scheme for `main`, but is in the
5.6 repository. You wouldn't know that though, since it would complain
about `repo_path` not being in the error message before crashing.

I've gone ahead and done some cleanups, reporting what the problem is at
the point of failure beyond the "key error 'swift-cmark-cfg'. Modernized
the error reporter a bit, so it should be a smidge easier to read.

etcwilde commented:
@swift-ci please test

etcwilde commented:
Pretty sure this one isn't me:
```
/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/swift/SwiftCompilerSources/Sources/Basic/Utils.swift:31:27: error: value of type 'llvm.StringRef' has no member '__bytes_beginUnsafe'
      start: lhs._bridged.__bytes_beginUnsafe(),
             ~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~~
/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/swift/SwiftCompilerSources/Sources/Basic/Utils.swift:32:66: error: value of type 'llvm.StringRef' has no member '__bytes_beginUnsafe'
      count: Int(lhs._bridged.__bytes_endUnsafe() - lhs._bridged.__bytes_beginUnsafe()))
                                                    ~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~~
```

etcwilde commented:
@swift-ci please test macOS platform

shahmishal commented:
@swift-ci python lint

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci please python lint

porglezomp commented on utils/update_checkout/update_checkout/update_checkout.py:
```diff
@@ -100,20 +115,25 @@ def get_branch_for_repo(config, repo_name, scheme_name, scheme_map,
+        except KeyError as e:
+            print(f"Failed to look up {repo_name} in scheme", file=sys.stderr)
+            raise e
+            exit(1)
```
If you want to exit, this should be `sys.exit` instead of `exit` (the latter is designed for interactive use).
But, should this just be removed since it's right after that `raise`?

porglezomp commented on utils/update_checkout/update_checkout/update_checkout.py:
```diff
@@ -100,20 +115,25 @@ def get_branch_for_repo(config, repo_name, scheme_name, scheme_map,
+                           .format(pr_id, repo_branch), "--tags"], echo=True)
+        except KeyError as e:
+            print(f"Failed to look up {repo_name} in scheme", file=sys.stderr)
+            raise e
```
Do you want `raise e` or just `raise`?
I believe using `raise e` will replace the call stack, and so might either be removing information, or just producing more verbose output if this ends up giving a chained error.

etcwilde replied:
Yeah, with the `raise e` its; 
```
Error on repo "/private/tmp/swift-test/swift-cmark-gfm": Traceback (most recent call last):
  File "/private/tmp/swift-test/swift/utils/update_checkout/update_checkout/update_checkout.py", line 156, in update_single_repository
    checkout_target, cross_repo = get_branch_for_repo(
  File "/private/tmp/swift-test/swift/utils/update_checkout/update_checkout/update_checkout.py", line 135, in get_branch_for_repo
   raise e
  File "/private/tmp/swift-test/swift/utils/update_checkout/update_checkout/update_checkout.py", line 119, in get_branch_for_repo
    scheme_branch = scheme_map[repo_name]
KeyError: 'swift-cmark-gfm'
```

Without it, it's just;
```
Error on repo "/private/tmp/swift-test/swift-cmark-gfm": Traceback (most recent call last):
  File "/private/tmp/swift-test/swift/utils/update_checkout/update_checkout/update_checkout.py", line 156, in update_single_repository
    checkout_target, cross_repo = get_branch_for_repo(
  File "/private/tmp/swift-test/swift/utils/update_checkout/update_checkout/update_checkout.py", line 119, in get_branch_for_repo
    scheme_branch = scheme_map[repo_name]
KeyError: 'swift-cmark-gfm'
```

Nice catch.

etcwilde committed:
Smol cleanups

Removing the `exit(1)` after failing. This shouldn't have an impact
since we've already failed and shouldn't have hit that. I may have kept
it from earlier debugging.

Replace `raise e` with just `raise`. It removes one level from the stack
trace where the actual `raise` takes place. The place we're actually
interested in is where the failure actually takes place.

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci python lint

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci please python lint

porglezomp commented on utils/update_checkout/update_checkout/update_checkout.py:
```diff
@@ -60,18 +60,33 @@ def check_parallel_results(results, op):
-            if r.stderr:
-                print(r.stderr)
-    return fail_count
+        try:
+            raise r
+        except KeyError as e:
+            messages.append(f"Failed to look up {e} in scheme\n")
+        except Exception:
```
I just realized you can make this a bit simpler and avoid the "`raise` for pattern matching" which is a bit quirky
```suggestion
        if isinstance(r, KeyError):
            messages.append(f"Failed to look up {r} in scheme\n")
        else:
```

etcwilde replied:
I wasn't sure which read better. It should probably be single `JobError` with a `__str__` method that just prints the right thing so this can just iterate over the list and print it without having to figure out what happened in the job that failed (even the KeyError here could lead to a misleading error message), but that is a bigger PR.

etcwilde commented:
@swift-ci please test

hnrklssn commented:
@etcwilde does the issue this fixes still exist?

etcwilde commented:
I believe the failure output is still not great. I should fix this back up and get it merged.
